lib/board.rb

```diff
@@ -2,21 +2,84 @@ class Board
 
   def initialize(side_length)
     @side_length = side_length
-    @board = Array.new(@side_length * @side_length, "")
+    @all_moves = set_all_moves(@side_length)
+    @player_moves = Hash.new
+  end
+
+  def to_string_array
+    string_array = Array.new(@all_moves.size, "")
+
+    @player_moves.each do |token, moves_set|
+      assign_moves_to_array(string_array, token, moves_set)
+    end
+
+    string_array
   end
 
   def move(move, player_token)
-    move_index = move - 1
+    if !@player_moves.has_key?(player_token)
+      @player_moves[player_token] = Set.new [move] 
+    else
+      @player_moves[player_token].add(move)
+    end
+  end
+
+  def get_available_moves
+    available_moves_set = @all_moves
 
-    @board[move_index] = player_token
+    @player_moves.each do |token, moves_set|
+      available_moves_set -= moves_set
+    end
+    
+    available_moves_set.to_a
   end
 
-  def to_string_array
-    @board
+  def set_board(board_string_array)
+    board_string_array.each_with_index do |token, index|
+      move(index + 1, token) if (token != "")
+    end
+  end
+
+  def get_side_length
+    @side_length
   end
 
-  def set_board(board)
-    @board = board
+  def get_move_set_for_player(token)
+    if @player_moves[token]
+      @player_moves[token]
+    else
+      [].to_set
```

Comment:
Why start with an array if you are just going to convert to a Set?

Reply:
Changed to Set.new instead